packages/backend/src/apps/m365-excel/common/file-privacy.ts

```diff
@@ -0,0 +1,227 @@
+    {
+      event: 'm365-queried-sharepoint-file-permissions',
+    },
+  )
+
+  // On M365, we're guaranteed that SharePoint's login name ends with the
+  // user's email. So this queries for permissions via a suffix filter on
+  // siteUser.loginName, which is a documented field.
+  //
+  // More info:
+  // - siteUser.loginName format:
+  //   https://learn.microsoft.com/en-us/sharepoint/dev/sp-add-ins/get-user-identity-and-properties-in-sharepoint#retrieve-current-website-user-identity-by-using-the-web-object
+  // - Schema:
+  //   https://learn.microsoft.com/en-us/graph/api/resources/sharepointidentity?view=graph-rest-1.0
+  const sharePointFilePermissionsResponse = await http.get(
+    `/v1.0/sites/${tenant.sharePointSiteId}/drive/items/${fileId}/permissions?$filter=endsWith(grantedToV2/siteUser/loginName,'${userEmail}')&$select=grantedToV2,roles`,
+  )
+  const permissions = sharepointFilePermissionsSchema.parse(
+    sharePointFilePermissionsResponse.data,
+  )
+
+  return permissions.some(
+    (p) => p.roles.includes('write') || p.roles.includes('owner'),
+  )
+}
+
+// https://learn.microsoft.com/en-us/graph/api/resources/permission?view=graph-rest-1.0
+//
+// NOTE: We're very loose with parsing this as we're _only_ interested in users,
+// but MS Graph may return other identities like app identities.
+const permissionUserSchema = z.object({
+  user: z
+    .object({
+      email: z.string().email().optional(),
+    })
+    .optional(),
+})
+const permissionSchema = z
+  .object({
+    roles: sharePointRolesSchema,
+    grantedTo: permissionUserSchema.optional(),
+    grantedToV2: permissionUserSchema.optional(),
+
+    // NOTE: we do not process grantedToIdentities nor grantedToIdentitiesV2
+    // because they are only applicable for shared links, not files.
+  })
+  .transform((permission) => ({
+    roles: permission.roles,
+    email:
+      permission.grantedToV2?.user?.email ??
+      permission.grantedTo?.user?.email ??
+      null,
+  }))
+
+const fileInfoResponseSchema = z
+  .object({
+    sensitivityLabel: z.object({
+      id: z.string().toUpperCase().nullish(),
+    }),
+    // https://learn.microsoft.com/en-us/graph/api/resources/itemreference?view=graph-rest-1.0
+    parentReference: z.object({
+      driveType: z.enum(['documentLibrary', 'personal', 'business']),
+      id: z.string().toUpperCase(),
+      siteId: z.string().toUpperCase().nullish(),
+    }),
+    permissions: z.array(permissionSchema),
+  })
+  .transform((response) => {
+    // EDGE CASE: Technically, the email field in grantedTo.user is undocumented
+    // and there is a very small chance that MS Graph will stop returning it. We
+    // only use it because it allows us to save 1 query per file access.
+    //
+    // To mitigate the risk of using this field, we want to detect if its is
+    // deprecated by Microsoft (and fall back to
+    // userHasWriteAccessAccordingToSharePointFilePermissionsFORBACKUPONLY).
```

Comment:
@pregnantboy do you think this should be in code or in oncall runbook?

I put it in code first because this has potential to break _all_ m365 excel pipes. But not sure if added complexity is worh it.